reader.go

```diff
@@ -29,31 +66,847 @@ type Reader struct {
+				MemberID: memberID,
+				MemberAssignments: groupAssignment{
+					Version: 1,
+					Topics:  topics,
+				}.bytes(),
+			})
+		}
+	}
+
+	return request
+}
+
+func (r *Reader) syncGroup(memberAssignments memberGroupAssignments) (map[string][]int32, error) {
+	conn, err := r.connect(r.stctx)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	request := r.newSyncGroupRequestV1(memberAssignments)
+	response, err := conn.syncGroups(request)
+	if err != nil {
+		switch err {
+		case UnknownMemberId:
+			r.mutex.Lock()
+			r.memberID = ""
+			r.mutex.Unlock()
+			return nil, err
+
+			// other possible return codes for sync group
+		//case GroupCoordinatorNotAvailable:
+		//case NotCoordinatorForGroup:
+		//case IllegalGeneration:
+		//case RebalanceInProgress:
+		//case GroupAuthorizationFailed:
+		default:
+			r.withLogger(func(l *log.Logger) {
+				l.Printf("joinGroup failed: %v\n", err)
+			})
+			return nil, err
+		}
+	}
+
+	assignments := groupAssignment{}
+	reader := bufio.NewReader(bytes.NewReader(response.MemberAssignments))
+	if _, err := (&assignments).readFrom(reader, len(response.MemberAssignments)); err != nil {
+		r.withErrorLogger(func(l *log.Logger) {
+			l.Printf("unable to read SyncGroup response for group, %v: %v\n", r.config.GroupID, err)
+		})
+		return nil, err
+	}
+
+	r.withLogger(func(l *log.Logger) {
+		l.Printf("sync group finished for group, %v\n", r.config.GroupID)
+	})
+
+	return assignments.Topics, nil
+}
+
+func (r *Reader) rebalance() (map[string][]int32, error) {
+	r.withLogger(func(l *log.Logger) {
+		l.Printf("rebalancing consumer group, %v", r.config.GroupID)
+	})
+
+	if err := r.refreshCoordinator(); err != nil {
+		return nil, err
+	}
+
+	members, err := r.joinGroup()
+	if err != nil {
+		return nil, err
+	}
+
+	subs, err := r.syncGroup(members)
+	if err != nil {
+		return nil, err
+	}
+
+	return subs, nil
+}
+
+func (r *Reader) unsubscribe() error {
+	r.cancel()
+	r.join.Wait()
+	return nil
+}
+
+func (r *Reader) fetchOffsets(subs map[string][]int32) (map[int]int64, error) {
+	conn, err := r.connect(r.stctx)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	partitions := subs[r.config.Topic]
+	offsets, err := conn.offsetFetch(offsetFetchRequestV3{
+		GroupID: r.config.GroupID,
+		Topics: []offsetFetchRequestV3Topic{
+			{
+				Topic:      r.config.Topic,
+				Partitions: partitions,
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	offsetsByPartition := map[int]int64{}
+	for _, pr := range offsets.Responses[0].PartitionResponses {
+		for _, partition := range partitions {
+			if partition == pr.Partition {
+				offset := pr.Offset
+				if offset >= 0 {
+					offset++ // advance to next offset
+				}
+				offsetsByPartition[int(partition)] = offset
+			}
+		}
+	}
+
+	return offsetsByPartition, nil
+}
+
+func (r *Reader) subscribe(subs map[string][]int32) error {
+	// always clear prior subscriptions
+	r.unsubscribe()
```

Comment:
It seems like `subscribe` is only called it one place, can we move this call out of this method so we keep a single responsibility per method?

Reply:
yep.  felt a little awkward to me as well, but I saw

```
r.cancel() // always cancel the previous reader
```
 
in start which seemed conceptually similar.